### PR TITLE
INTYGFV-16064: Implemented GetCertificatePrintService that returns Certificate in PDF-format.

### PR DESCRIPTION
**Pull request by ChrilleSigerhed**

### CODE REVIEW

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]

File: web/src/main/java/se/inera/intyg/webcert/web/service/facade/internalapi/service/GetCertificatePrintService.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import se.inera.intyg.common.support.modules.registry.ModuleNotFoundException;
+import se.inera.intyg.common.support.modules.support.ApplicationOrigin;
+import se.inera.intyg.common.support.modules.support.api.ModuleApi;
+import se.inera.intyg.common.support.modules.support.api.dto.PdfResponse;
+import se.inera.intyg.common.support.modules.support.api.exception.ModuleException;
+import se.inera.intyg.webcert.web.converter.util.IntygConverterUtil;
+import se.inera.intyg.webcert.web.web.controller.internalapi.dto.GetUtkastResponse;
+
+@Service
+public class GetCertificatePrintService {
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply web/src/main/java/se/inera/intyg/webcert/web/service/facade/internalapi/service/GetCertificatePrintService.java:38]: Changed it to GetCertificatePdfService.

File: web/src/main/java/se/inera/intyg/webcert/web/service/facade/internalapi/service/GetCertificatePrintService.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    private static final String DONT_DISPLAY_DIAGNOSIS_ID = "!diagnoser";
+    private static final List<String> OPTIONAL_FIELDS = List.of(DONT_DISPLAY_DIAGNOSIS_ID, "!onskarFormedlaDiagnoser");
+
+    public GetCertificatePrintService(GetUtkastFacadeService getUtkastFacadeService, IntygModuleRegistry moduleRegistry) {
+        this.getUtkastFacadeService = getUtkastFacadeService;
+        this.moduleRegistry = moduleRegistry;
+    }
+
+    public PdfResponse get(String customizationId, Certificate certificate, boolean pdlLog,
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply web/src/main/java/se/inera/intyg/webcert/web/service/facade/internalapi/service/GetCertificatePrintService.java:51]: Moved them inside service.

File: web/src/main/java/se/inera/intyg/webcert/web/service/facade/internalapi/service/GetCertificatePrintService.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        boolean validateAccess)
+        throws ModuleNotFoundException, ModuleException {
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply web/src/main/java/se/inera/intyg/webcert/web/service/facade/internalapi/service/GetCertificatePrintService.java:53]: Added exception management

File: web/src/main/java/se/inera/intyg/webcert/web/service/facade/internalapi/service/GetCertificatePrintService.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    private void validateCertificate(Certificate certificate) {
+        if (certificate.getMetadata().getStatus().equals(CertificateStatus.REVOKED)) {
+            throw new IllegalStateException("Revoked certificate can not be printed");
+        }
+    }
+
+    private PdfResponse getPdfResponse(GetUtkastResponse utkastResponse, String customizationId, ModuleApi moduleApi, String jsonModel,
+        List<Status> statuses) throws ModuleException {
+        if (customizationId != null && customizationId.equals(DONT_DISPLAY_DIAGNOSIS_ID)) {
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply web/src/main/java/se/inera/intyg/webcert/web/service/facade/internalapi/service/GetCertificatePrintService.java:73]: Done!

File: web/src/main/java/se/inera/intyg/webcert/web/service/facade/internalapi/service/GetUtkastFacadeService.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+import se.inera.intyg.webcert.persistence.utkast.model.Utkast;
+import se.inera.intyg.webcert.web.integration.ITIntegrationService;
+import se.inera.intyg.webcert.web.service.access.DraftAccessServiceHelper;
+import se.inera.intyg.webcert.web.service.facade.CertificateTextVersionFacadeService;
+import se.inera.intyg.webcert.web.service.intyg.IntygService;
+import se.inera.intyg.webcert.web.service.utkast.UtkastService;
+import se.inera.intyg.webcert.web.web.controller.internalapi.dto.GetUtkastResponse;
+
+@Service
+public class GetUtkastFacadeService {
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply web/src/main/java/se/inera/intyg/webcert/web/service/facade/internalapi/service/GetUtkastFacadeService.java:39]: Changed it to GetRequiredFieldsForCertificatePdfService

File: web/src/main/java/se/inera/intyg/webcert/web/web/controller/internalapi/CertificateInternalApiController.java [hunk 6]
```diff
@@ -60,4 +71,25 @@ public GetCertificateResponse getCertificate(@PathParam("certificateId") String
             availableFunction
         );
     }
+
+    @POST
+    @Path("/pdf/{certificateId}")
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply web/src/main/java/se/inera/intyg/webcert/web/web/controller/internalapi/CertificateInternalApiController.java:76]: Updated.

File: web/src/main/java/se/inera/intyg/webcert/web/web/controller/internalapi/CertificateInternalApiController.java [hunk 7]
```diff
@@ -60,4 +71,25 @@ public GetCertificateResponse getCertificate(@PathParam("certificateId") String
...
+    @PrometheusTimeMethod
+    @Produces(MediaType.APPLICATION_JSON + UTF_8_CHARSET)
+    @Consumes(MediaType.APPLICATION_JSON + UTF_8_CHARSET)
+    public PrintCertificateResponseDTO getPdfData(@RequestBody PrintCertificateRequestDTO request,
+        @PathParam("certificateId") String certificateId)
+        throws ModuleNotFoundException, ModuleException {
+        final var certificate = getCertificateFacadeService.getCertificate(certificateId, SHOULD_NOT_PDL_LOG, SHOULD_NOT_VALIDATE_ACCESS);
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply web/src/main/java/se/inera/intyg/webcert/web/web/controller/internalapi/CertificateInternalApiController.java:83]: No it was not necessary, good catch! Updated

File: web/src/main/java/se/inera/intyg/webcert/web/web/controller/internalapi/dto/CertificatePdfResponseDTO.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ */
+
+package se.inera.intyg.webcert.web.web.controller.internalapi.dto;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+public class CertificatePdfResponseDTO {
+
+    String filename;
```
Comment (mww-sopra-steria) [inline web/src/main/java/se/inera/intyg/webcert/web/web/controller/internalapi/dto/CertificatePdfResponseDTO.java:27]: Set as private

**Review by mww-sopra-steria — APPROVED** [review]
Looks great!

**Pull request merged**